scipy/integrate/_cubature.py

```diff
@@ -258,13 +282,46 @@ def cubature(f, a, b, *, rule="gk21", rtol=1e-8, atol=0, max_subdivisions=10000,
     >>> res.estimate
      array([[-0.79812452,  0.35246913, -0.52273628],
             [ 0.88392779,  0.59139899,  0.41895111]])
+
+    **2D integral with infinite limits**:
+
+    .. math::
+
+        \int^{ \infty }_{ -\infty }
+        \int^{ \infty }_{ -\infty }
+            e^{-x^2-y^2}
+        \text dy
+        \text dx
+
+    >>> def gaussian(x):
+    ...     return np.exp(-np.sum(x**2, axis=-1))
+    >>> res = cubature(gaussian, [-np.inf, -np.inf], [np.inf, np.inf])
+    >>> res.estimate
+     3.1415926
+
+    **1D integral with singularities avoided using** `points`:
+
+    .. math::
+
+        \int^{ 1 }_{ -1 }
+          \frac{\sin(x)}{x}
+        \text dx
+
+    It is necessary to use the `points` parameter to avoid evaluating `f` at the origin.
+
+    >>> def sinc(x):
+    ...     return np.sin(x)/x
+    >>> res = cubature(sinc, [-1], [1], points=[[0]])
+    >>> res.estimate
+     1.8921661
     """
 
     # It is also possible to use a custom rule, but this is not yet part of the public
     # API. An example of this can be found in the class scipy.integrate._rules.Rule.
 
     xp = array_namespace(a, b)
     max_subdivisions = float("inf") if max_subdivisions is None else max_subdivisions
+    points = [] if points is None else [xp.asarray(p, dtype=xp.float64) for p in points]
```

Comment:
Is there a reason for not using the default floating type? or maybe something using `xp.result_type`

Reply:
Thanks for the review. I'm a bit confused here, would the "default floating type" be something like `dtype=xp.floating` or `dtype=float`? I'm applying `xp.asarray` to each element in the list so that `points` can be specified as `array_like`. If using `dtype=xp.float64` is a problem here, is it also a problem on the lines below this one?

```
# Convert a and b to arrays
a = xp.asarray(a, dtype=xp.float64)
b = xp.asarray(b, dtype=xp.float64)
```

Reply:
Sorry this was a slightly rushed comment. What I meant was this forces everything to be `xp.float64` whereas in some libraries (pytorch) the default floating dtype is `xp.float32`. If we did something like `a = xp.asarray(a, dtype=xp.results_type(a, b, 1.0))` then we don't promote things unnecessarily. cc @lucascolley @mdhaber who can probably explain more coherently than me

Reply:
I think `_array_api.xp_broadcast_promote` with `force_floating=True` may be useful here.
https://github.com/scipy/scipy/blob/ef157bf12f15cbf47f04f2d38576499be5a7703b/scipy/_lib/_array_api.py#L587

Reply:
It would also be good to add a test for different `dtype`s of `a` and `b` to check the `dtype` of the result is promoted correctly

Reply:
Just writing some tests for this and found the following case: Rules like `gk21` return the nodes and weights as arrays using `xp.asarray([0.996, ..., -0.996])` and so are using the default floating dtype in that array namespace.

Since the default floating dtype in Numpy is `float64`, this means that if a user does something like:

```
cubature(
        f,
        np.asarray([0], dtype=np.float32),
        np.asarray([1], dtype=np.float32),
)
```

then the result will actually have dtype `float64` rather than `float32` because calculations are done with the nodes and weights, which themselves will have dtype `float64`. If they were using `torch` instead, then the result would have dtype `float32`. Is this what should happen?

If the nodes and weights should also be returned with dtype `float32` in the NumPy case, then I think it would be necessary to add a new parameter like `result_dtype` to the constructor of each `FixedRule`, since each `Rule` can be initialised independently without `a` or `b`.

Reply:
> Is this what should happen?

I think the whole calculation should respect the common dtype of the user input. 

> Rules like gk21 return the nodes and weights as arrays using `xp.asarray([0.996, ..., -0.996])` and so are using the default floating dtype in that array namespace.

It sounds like this would mean that even if the user were to provide `float64` input, the _actual_ precision of the output is no better than the default dtype of the array library. Is that right?

I'd suggest storing the nodes and weights with `float64` and downcasting if necessary before use (this is what I do, anyway) or adding a `dtype` parameter.